contentreports-filtereditems.md

```diff
@@ -0,0 +1,129 @@
+# Displaying the filtered collections report
+[Back to the list of all defined endpoints](endpoints.md)
+
+## Statistics for the whole repository
+**POST /api/contentreports/filtereditems**
```

Comment:
Why is this a POST instead of a GET?  I notice that the "statistics" endpoints always use GET _except when they are adding data to the statistics_.  See https://github.com/DSpace/RestContract/blob/main/statistics-reports.md and https://github.com/DSpace/RestContract/blob/main/statistics-viewevents.md

Could we better describe why we need to use POST for these endpoints?  It appears they are readonly, which implies they might be switched to GET.

Reply:
That's true, I thought about it. My concern with GET is, as you suggested above, the limited length of the parameters passed as part of the URL. There should be no problems with the Filtered Collections report (only Boolean filters).

Parameterization of the Filtered Items report, however, is much more complex and can easily become long enough to exceed any limit enforced by application servers for URL query strings. This is why I implemented this report as a POST endpoint.

For (a bit of) uniformity, I also added POST support to the Filtered Collections report.

Besides, while the HTTP spec clearly states that GET should be used for read-only requests, I saw nothing stating that POST should be used only for data changing requests.

If you feel that everything should be switched to GET anyway, I can do it. In this case, the Filtered Items report shall be thoroughly tested with lots of parameters selected to make sure that nothing goes wrong due to too long a query string.

Reply:
About Filtered Items: The parameters could be organized into a GET query string, although such a string might become quite long. Another concern I had while designing the API for this report is the "query predicates" part: these are structured parameters (a query predicate is a (field, operator, value) tuple). This is why I didn't include a GET version of this report.

Reply:
Please check: https://github.com/DSpace/RestContract/blob/main/search-endpoint.md solution.

A query string solution it's used for filtering results:
```
f.<:filter-name>=<:filter-value>,<:filter-operator>
```

where a `filter-name` belongs to a predefined structure that is previously returned. Like: `f.title=rainbows,notcontains`

``` json
{
  "query":"my query",
  "scope":"9076bd16-e69a-48d6-9e41-0238cb40d863",
  "appliedFilters": [
      {
        "filter" : "title",
        "operator" : "notcontains",
        "value" : "abcd",
        "label" : "abcd"
      },
```

Reply:
I think this discussion stems from a disagreement between HTTP and REST about what POST is for.  RFC 9110 says that creating a new resource is only one possible use for POST.  "The POST method requests that the target resource process the representation enclosed in the request according to the resource's own specific semantics."  The description of POST [here](https://restfulapi.net/rest-put-vs-post/) is quite a bit narrower.  One might say that REST is a *reuse* of HTTP syntax with different semantics.

So it can be argued that REST is not a very good fit for this operation, but it is what we have.

Would it be a violation of the "spirit of REST" to consider such a POST to be storing a report description, which is consumed in the act of generating the report?  Reports may take some time to create.  Suppose that one POSTs a document describing the desired report and receives a token in the response.  The report generator runs in the background.  When finished, the token can be presented (using GET) to receive the report, and the report description is then destroyed.